pennylane/io/qualtran_io.py

```diff
@@ -72,6 +76,313 @@ def _(op: qtemps.subroutines.qpe.QuantumPhaseEstimation):
+                clean=False,
+            )
+
+            gate_types[_map_to_bloq()(qrom_op)] += 1
+            gate_types[_map_to_bloq()(qops.adjoint(qrom_op))] += 1
+
+        gate_types[_map_to_bloq()(qops.CRY(0, wires=[0, 1]))] = (
+            num_precision_wires * num_state_qubits
+        )
+
+        if not positive_and_real:
+            num_bit_flips = 2 ** (num_state_qubits - 1)
+            zero_string = "0" * num_precision_wires
+            one_string = "0" * (num_precision_wires - 1) + "1" if num_precision_wires > 0 else ""
+            bitstrings = [zero_string for _ in range(num_bit_flips)] + [
+                one_string for _ in range(num_bit_flips)
+            ]
+
+            qrom_op = qtemps.QROM(
+                bitstrings=bitstrings,
+                target_wires=precision_wires,
+                control_wires=input_wires,
+                work_wires=work_wires,
+                clean=False,
+            )
+
+            gate_types[_map_to_bloq()(qrom_op)] += 1
+            gate_types[_map_to_bloq()(qops.adjoint(qrom_op))] += 1
+            gate_types[
+                _map_to_bloq()(
+                    qops.ctrl(
+                        qops.GlobalPhase((2 * np.pi), wires=input_wires[0]),
+                        control=0,
+                    )
+                )
+            ] = num_precision_wires
+
+        return gate_types
+
+    @_op_call_graph.register
+    def _(op: qops.BasisState):
+        gate_types = {}
+        gate_types[qt_gates.XGate()] = sum(op.parameters[0])
+
+        return gate_types
+
+    @_op_call_graph.register
+    def _(op: qtemps.subroutines.QROM):
+        gate_types = defaultdict(int)
+        bitstrings = op.hyperparameters["bitstrings"]
+        num_bitstrings = len(bitstrings)
+
+        num_bit_flips = 0
+        for bit_string in bitstrings:
+            num_bit_flips += bit_string.count("1")
+
+        num_work_wires = len(op.hyperparameters["work_wires"])
+        size_bitstring = len(op.hyperparameters["target_wires"])
+        num_control_wires = len(op.hyperparameters["control_wires"])
+        clean = op.hyperparameters["clean"]
+
+        if num_control_wires == 0:
+            gate_types[qt_gates.XGate()] = num_bit_flips
+            return gate_types
+
+        cnot = qt_gates.CNOT()
+        hadamard = qt_gates.Hadamard()
+        num_parallel_computations = (num_work_wires + size_bitstring) // size_bitstring
+        # num_parallel_computations = min(num_parallel_computations, num_bitstrings)
+
+        square_fact = math.floor(
+            math.sqrt(num_bitstrings)
+        )  # use a square scheme for rows and cloumns
+        num_parallel_computations = min(num_parallel_computations, square_fact)
+
+        num_swap_wires = math.floor(math.log2(num_parallel_computations))
+        num_select_wires = math.ceil(math.log2(math.ceil(num_bitstrings / (2**num_swap_wires))))
+        assert num_swap_wires + num_select_wires <= num_control_wires
```

Comment:
Do we need this `assert`? Can we raise a `ValueError` instead?

Reply:
I can just delete it, it's copy and pasted from the resources.